# Objetivo

Encontrar uma configuração para uma rede neural Keras que atinja mais de 90% de precisão classificando amostras da base de dados Breast Cancer Winsconsin.<br>
<br>
O dataset utilizado é o Breast Cancer Winsconsin, do repositório UC Irvine.<br>
Possui 569 amostras, cada uma com 32 variáveis (1 ID, 1 target e 30 atributos).<br>
Target: 0 para begnino, 1 para maligno (classificação binária).<br>
Não há dados faltantes.<br>

# Importação dos recursos

In [3]:
import pandas as pd
import tensorflow.keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, InputLayer
from scikeras.wrappers import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

2024-08-22 11:20:04.472345: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [4]:
previsores = pd.read_csv('Deep Learning com Python de A a Z/Parte 1 - Redes Neurais Artificiais/classificação binária/entradas_breast.csv')
classe = pd.read_csv('Deep Learning com Python de A a Z/Parte 1 - Redes Neurais Artificiais/classificação binária/saidas_breast.csv')

# Definição da rede neural

In [5]:
def criarRede(neurons, activation, kernel_initializer, optimizer, loss):
    classificador = Sequential()
    classificador.add(InputLayer(shape=(30,)))
    classificador.add(Dense(units = neurons, activation = activation, 
                        kernel_initializer = kernel_initializer))
    classificador.add(Dropout(0.2))
    classificador.add(Dense(units = neurons, activation = activation, 
                        kernel_initializer = kernel_initializer))
    classificador.add(Dropout(0.2))
    classificador.add(Dense(units = 1, activation = 'sigmoid'))
    classificador.compile(optimizer = optimizer, loss = loss,
                      metrics = ['binary_accuracy'])
    return classificador

In [6]:
classificador = KerasClassifier(model = criarRede)

# Testando Batch Size

In [10]:
batchTest = {
    'batch_size': [10, 30],
    'epochs': [50],
    'model__neurons': [8],
    'model__activation': ['relu'],
    'model__kernel_initializer': ['random_uniform'],
    'model__optimizer': ['adam'],
    'model__loss': ['binary_crossentropy']
}

In [11]:
grid_search_batch = GridSearchCV(estimator = classificador, param_grid=batchTest, scoring = 'accuracy', 
                                 cv=10)

In [13]:
grid_search_batch = grid_search_batch.fit(previsores, classe)

Epoch 1/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - binary_accuracy: 0.5697 - loss: 1.0875
Epoch 2/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6042 - loss: 0.6800
Epoch 3/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.6975 - loss: 0.5806
Epoch 4/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7130 - loss: 0.5561
Epoch 5/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6838 - loss: 0.5170
Epoch 6/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7595 - loss: 0.4909
Epoch 7/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.7161 - loss: 0.4938
Epoch 8/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7605 - loss: 0.4790
Epoch 9/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7889 - loss: 0.4902
Epoch 10/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.7594 - loss: 0.4657
Epoch 11/50
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.8322 - loss: 0.43

In [15]:
grid_search_batch.best_params_

{'batch_size': 10,
 'epochs': 50,
 'model__activation': 'relu',
 'model__kernel_initializer': 'random_uniform',
 'model__loss': 'binary_crossentropy',
 'model__neurons': 8,
 'model__optimizer': 'adam'}

Batch size de 10 parece ser o melhor que o de 50.

In [16]:
grid_search_batch.best_score_

0.8946115288220551

# Testando número de epochs

In [17]:
epochs_test = {
    'batch_size': [10],
    'epochs': [100, 200],
    'model__neurons': [8],
    'model__activation': ['relu'],
    'model__kernel_initializer': ['random_uniform'],
    'model__optimizer': ['adam'],
    'model__loss': ['binary_crossentropy']
}

In [18]:
grid_search_epochs = GridSearchCV(estimator = classificador, param_grid=epochs_test, 
                                  scoring = 'accuracy', cv = 10)

In [19]:
grid_search_epochs = grid_search_epochs.fit(previsores, classe)

Epoch 1/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - binary_accuracy: 0.5704 - loss: 0.8945
Epoch 2/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.6355 - loss: 0.6546
Epoch 3/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.6988 - loss: 0.5914
Epoch 4/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7091 - loss: 0.5410
Epoch 5/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7895 - loss: 0.5050
Epoch 6/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7820 - loss: 0.4883
Epoch 7/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7679 - loss: 0.4612
Epoch 8/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.7926 - loss: 0.4569
Epoch 9/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - binary_accuracy: 0.8073 - loss: 0.4277
Epoch 10/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - binary_accuracy: 0.8272 - loss: 0.4045
Epoch 11/100
52/52 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - binary_accuracy: 0.7971 -

In [20]:
grid_search_epochs.best_params_

{'batch_size': 10,
 'epochs': 100,
 'model__activation': 'relu',
 'model__kernel_initializer': 'random_uniform',
 'model__loss': 'binary_crossentropy',
 'model__neurons': 8,
 'model__optimizer': 'adam'}

In [21]:
grid_search_epochs.best_score_

0.9121553884711779

# Conclusão

Como demonstrado pelas Grid Searches acima, uma rede com as seguintes configurações atinge 91% de precisão:

{'batch_size': 10,
 'epochs': 100,
 'model__activation': 'relu',
 'model__kernel_initializer': 'random_uniform',
 'model__loss': 'binary_crossentropy',
 'model__neurons': 8,
 'model__optimizer': 'adam'}